test/lib/tpunit++.hpp

```diff
@@ -75,22 +76,22 @@ using namespace std;
  * executing test function to continue if the supplied predicate
  * is not satisified.
  */
-#define ASSERT_TRUE(condition) if(condition) { PASS(); } else { ABORT(); }
-#define EXPECT_TRUE(condition) if(condition) { PASS(); } else { FAIL(); }
-#define ASSERT_FALSE(condition) if(condition) { ABORT(); } else { PASS(); }
-#define EXPECT_FALSE(condition) if(condition) { FAIL(); } else { PASS(); }
-#define ASSERT_EQUAL(lhs, rhs) if((lhs) == (rhs)) { PASS(); } else { ABORT(); }
-#define EXPECT_EQUAL(lhs, rhs) if((lhs) == (rhs)) { PASS(); } else { FAIL(); }
-#define ASSERT_NOT_EQUAL(lhs, rhs) if((lhs) != (rhs)) { PASS(); } else { ABORT(); }
-#define EXPECT_NOT_EQUAL(lhs, rhs) if((lhs) != (rhs)) { PASS(); } else { FAIL(); }
-#define ASSERT_GREATER_THAN(lhs, rhs) if((lhs) > (rhs)) { PASS(); } else { ABORT(); }
-#define EXPECT_GREATER_THAN(lhs, rhs) if((lhs) > (rhs)) { PASS(); } else { FAIL(); }
-#define ASSERT_GREATER_THAN_EQUAL(lhs, rhs) if((lhs) >= (rhs)) { PASS(); } else { ABORT(); }
-#define EXPECT_GREATER_THAN_EQUAL(lhs, rhs) if((lhs) >= (rhs)) { PASS(); } else { FAIL(); }
-#define ASSERT_LESS_THAN(lhs, rhs) if((lhs) < (rhs)) { PASS(); } else { ABORT(); }
-#define EXPECT_LESS_THAN(lhs, rhs) if((lhs) < (rhs)) { PASS(); } else { FAIL(); }
-#define ASSERT_LESS_THAN_EQUAL(lhs, rhs) if((lhs) <= (rhs)) { PASS(); } else { ABORT(); }
-#define EXPECT_LESS_THAN_EQUAL(lhs, rhs) if((lhs) <= (rhs)) { PASS(); } else { FAIL(); }
+#define ASSERT_TRUE(condition) if(condition) { PASS(); } else { testLog("assertion failed: " + SToStr(#condition) + " resolved to FALSE"); ABORT(); }
+#define EXPECT_TRUE(condition) if(condition) { PASS(); } else { testLog("expect failed: " + SToStr(#condition) + " resolved to FALSE"); FAIL(); }
+#define ASSERT_FALSE(condition) if(condition) { testLog("assertion failed: " + SToStr(#condition) + " resolved to TRUE"); ABORT(); } else { PASS(); }
+#define EXPECT_FALSE(condition) if(condition) { testLog("expect failed: " + SToStr(#condition) + " resolved to TRUE"); FAIL(); } else { PASS(); }
+#define ASSERT_EQUAL(lhs, rhs) if((lhs) == (rhs)) { PASS(); } else { EqualComparator(lhs, rhs); ABORT(); }
```

Comment:
It occurs to me we don't actually need the comparison in `EqualComparator`, all we want is the generic printing that falls back to "unprintable" instead of failing to compile. The actual comparison has already been done here.

We could probably simplify all the template cone that way.

Reply:
Oh true, so we type-cast with the class first then print?

Reply:
Actually not sure what you mean here.

Reply:
We already know (for this line) that if we hit the `else`, these are not equal (otherwise we would have run `PASS()`).

So where we do `EqualComparator(lhs, rhs);` we could just do something like: `cout << "Not equal " << Print(lhs, rhs);` where `Print` was a simpler version of `EqualComparator` that didn't need to actually do any comparison, just the printing.

If that was the case we could replace:
```
        template <typename T, enable_if_t<is_integral<T>::value, bool> = true, typename U>
        EqualComparator(T a, U b)
        {
            // Integral case.
            if (Equal<T>(a) != (T) b) {
                cout << "(Integral): " << a << " != " << b << "\n";
            } else {
                cout << "(Integral): " << a << " == " << b << "\n";
            }
        }
```


With:
```
        template <typename T, enable_if_t<is_integral<T>::value, bool> = true, typename U>
        EqualComparator(T a, U b)
        {
            cout << "(Integral): " << a << " == " << b << "\n";
        }
```

And then we don't even need `class Equal` anymore.